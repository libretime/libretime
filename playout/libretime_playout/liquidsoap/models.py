from enum import Enum

from pydantic import BaseModel


class Info(BaseModel):
    station_name: str


class MessageFormatKind(int, Enum):
    ARTIST_TITLE = 0
    SHOW_ARTIST_TITLE = 1
    RADIO_SHOW = 2


class StreamPreferences(BaseModel):
    input_fade_transition: float
    message_format: MessageFormatKind
    message_offline: str
    replay_gain_enabled: bool
    replay_gain_offset: float


class StreamState(BaseModel):
    input_main_connected: bool
    input_main_streaming: bool
    input_show_connected: bool
    input_show_streaming: bool
    schedule_streaming: bool
