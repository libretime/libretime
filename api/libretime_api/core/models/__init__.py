from .auth import LoginAttempt, UserToken
from .preference import Preference
from .role import Role
from .service import ServiceRegister
from .user import User, UserManager
