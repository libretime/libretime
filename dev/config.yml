general:
  public_url: http://localhost:8080
  api_key: some_secret_api_key
  secret_key: some_secret_key

database:
  host: postgres

rabbitmq:
  host: rabbitmq

email:
  host: mailpit
  port: 1025

playout:
  liquidsoap_host: liquidsoap

liquidsoap:
  server_listen_address: 0.0.0.0

  harbor_ssl_certificate: /certs/fake.crt
  harbor_ssl_private_key: /certs/fake.key

stream:
  inputs:
    main:
      public_url: https://localhost:8001/main
      mount: main
      port: 8001
      secure: true

  outputs:
    .default_icecast_output: &default_icecast_output
      host: icecast
      port: 8000
      source_password: hackme
      admin_password: hackme
      name: LibreTime!
      description: LibreTime Radio!
      website: https://libretime.org
      genre: various

    icecast:
      - <<: *default_icecast_output
        enabled: true
        mount: main.ogg
        public_url: https://localhost:8443/main.ogg
        audio:
          format: ogg
          bitrate: 256

      - <<: *default_icecast_output
        enabled: true
        mount: main.opus
        public_url: https://localhost:8443/main.opus
        audio:
          format: opus
          bitrate: 256

      - <<: *default_icecast_output
        enabled: true
        mount: main.mp3
        public_url: https://localhost:8443/main.mp3
        audio:
          format: mp3
          bitrate: 256

    system:
      - enabled: false
        kind: pulseaudio
