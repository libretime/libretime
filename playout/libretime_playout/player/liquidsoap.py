import logging
import time
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Set

from ..liquidsoap.client import LiquidsoapClient
from ..utils import seconds_between
from .events import ActionEvent, AnyEvent, EventKind, FileEvent, WebStreamEvent

logger = logging.getLogger(__name__)


def create_liquidsoap_annotation(file_event: FileEvent) -> str:
    # We need liq_start_next value in the annotation. That is the value that controls
    # overlap duration of crossfade.
    annotations = {
        "media_id": file_event.id,
        "schedule_table_id": file_event.row_id,
        "liq_start_next": "0",
        "liq_fade_in": file_event.fade_in / 1000,
        "liq_fade_out": file_event.fade_out / 1000,
        "liq_cue_in": file_event.cue_in,
        "liq_cue_out": file_event.cue_out,
    }

    if file_event.replay_gain is not None:
        annotations["replay_gain"] = f"{file_event.replay_gain} dB"

    # Override the the artist/title that Liquidsoap extracts from a file's metadata with
    # the metadata we get from LibreTime. (You can modify metadata in LibreTime's library,
    # which doesn't get saved back to the file.)
    if file_event.artist_name:
        annotations["artist"] = file_event.artist_name.replace('"', '\\"')

    if file_event.track_title:
        annotations["title"] = file_event.track_title.replace('"', '\\"')

    annotations_str = ",".join(f'{key}="{value}"' for key, value in annotations.items()).replace("\n", "")

    return "annotate:" + annotations_str + ":" + str(file_event.local_filepath)


class TelnetLiquidsoap:
    current_prebuffering_stream_id: Optional[int] = None

    def __init__(
        self,
        liq_client: LiquidsoapClient,
        queues: List[int],
    ):
        self.liq_client = liq_client
        self.queues = queues

    def queue_clear_all(self):
        try:
            self.liq_client.queues_remove(*self.queues)
        except OSError as exception:
            logger.exception(exception)

    def queue_remove(self, queue_id: int):
        try:
            self.liq_client.queues_remove(queue_id)
        except OSError as exception:
            logger.exception(exception)

    def queue_push(self, queue_id: int, file_event: FileEvent):
        try:
            annotation = create_liquidsoap_annotation(file_event)
            self.liq_client.queue_push(queue_id, annotation, file_event.show_name)
        except OSError as exception:
            logger.exception(exception)

    def stop_web_stream_buffer(self):
        try:
            self.liq_client.web_stream_stop_buffer()
        except OSError as exception:
            logger.exception(exception)

    def stop_web_stream_output(self):
        try:
            self.liq_client.web_stream_stop()
        except OSError as exception:
            logger.exception(exception)

    def start_web_stream(self):
        try:
            self.liq_client.web_stream_start()
            self.current_prebuffering_stream_id = None
        except OSError as exception:
            logger.exception(exception)

    def start_web_stream_buffer(self, event: WebStreamEvent):
        try:
            self.liq_client.web_stream_start_buffer(event.row_id, event.uri)
            self.current_prebuffering_stream_id = event.row_id
        except OSError as exception:
            logger.exception(exception)

    def get_current_stream_id(self) -> str:
        try:
            return self.liq_client.web_stream_get_id()
        except OSError as exception:
            logger.exception(exception)
            return "-1"

    def disconnect_source(self, sourcename):
        if sourcename not in ("master_dj", "live_dj"):
            raise ValueError(f"invalid source name: {sourcename}")

        try:
            logger.debug("Disconnecting source: %s", sourcename)
            self.liq_client.source_disconnect(sourcename)
        except OSError as exception:
            logger.exception(exception)

    def switch_source(self, sourcename, status):
        if sourcename not in ("master_dj", "live_dj", "scheduled_play"):
            raise ValueError(f"invalid source name: {sourcename}")

        try:
            logger.debug('Switching source: %s to "%s" status', sourcename, status)
            self.liq_client.source_switch_status(sourcename, status == "on")
        except OSError as exception:
            logger.exception(exception)


class Liquidsoap:
    def __init__(self, liq_client: LiquidsoapClient):
        self.liq_queue_tracker: Dict[int, Optional[FileEvent]] = {
            0: None,
            1: None,
            2: None,
            3: None,
        }

        self.liq_client = liq_client
        self.telnet_liquidsoap = TelnetLiquidsoap(
            liq_client,
            list(self.liq_queue_tracker.keys()),
        )

    def play(self, event: AnyEvent) -> None:
        if isinstance(event, FileEvent):
            self.handle_file_type(event)
        elif isinstance(event, ActionEvent):
            self.handle_event_type(event)
        elif isinstance(event, WebStreamEvent):
            self.handle_web_stream_type(event)
        else:
            raise UnknownEvent(str(event))

    def handle_file_type(self, file_event: FileEvent) -> None:
        """
        Wait 200 seconds (2000 iterations) for file to become ready,
        otherwise give up on it.
        """
        iter_num = 0
        while not file_event.file_ready and iter_num < 2000:
            time.sleep(0.1)
            iter_num += 1

        if file_event.file_ready:
            available_queue = self.find_available_queue()

            try:
                self.telnet_liquidsoap.queue_push(available_queue, file_event)
                self.liq_queue_tracker[available_queue] = file_event
            except Exception as exception:
                logger.exception(exception)
                raise exception
        else:
            logger.warning(
                "File %s did not become ready in less than 5 seconds. Skipping...",
                file_event.local_filepath,
            )

    def handle_web_stream_type(self, event: WebStreamEvent) -> None:
        if event.type == EventKind.WEB_STREAM_BUFFER_START:
            self.telnet_liquidsoap.start_web_stream_buffer(event)
        elif event.type == EventKind.WEB_STREAM_OUTPUT_START:
            if event.row_id != self.telnet_liquidsoap.current_prebuffering_stream_id:
                # this is called if the stream wasn't scheduled sufficiently ahead of
                # time so that the prebuffering stage could take effect. Let's do the
                # prebuffering now.
                self.telnet_liquidsoap.start_web_stream_buffer(event)
            self.telnet_liquidsoap.start_web_stream()
        elif event.type == EventKind.WEB_STREAM_BUFFER_END:
            self.telnet_liquidsoap.stop_web_stream_buffer()
        elif event.type == EventKind.WEB_STREAM_OUTPUT_END:
            self.telnet_liquidsoap.stop_web_stream_output()

    def handle_event_type(self, event: ActionEvent) -> None:
        if event.event_type == "kick_out":
            self.telnet_liquidsoap.disconnect_source("live_dj")
        elif event.event_type == "switch_off":
            self.telnet_liquidsoap.switch_source("live_dj", "off")

    def find_available_queue(self) -> int:
        available_queue = None
        for queue_id, item in self.liq_queue_tracker.items():
            if item is None or item.ended():
                # queue "i" is available. Push to this queue
                available_queue = queue_id

        if available_queue is None:
            raise NoQueueAvailableException()

        return available_queue

    # pylint: disable=too-many-branches
    def verify_correct_present_media(self, scheduled_now: List[AnyEvent]) -> None:
        """
        verify whether Liquidsoap is currently playing the correct files.
        if we find an item that Liquidsoap is not playing, then push it
        into one of Liquidsoap's queues. If Liquidsoap is already playing
        it do nothing. If Liquidsoap is playing a track that isn't in
        currently_playing then stop it.

        Check for Liquidsoap media we should source.skip
        get liquidsoap items for each queue. Since each queue can only have one
        item, we should have a max of 8 items.

        2013-03-21-22-56-00_0: {
        id: 1,
        type: "stream_output_start",
        row_id: 41,
        uri: "http://stream2.radioblackout.org:80/blackout.ogg",
        start: "2013-03-21-22-56-00",
        end: "2013-03-21-23-26-00",
        show_name: "Untitled Show"
        },
        """

        scheduled_now_files: List[FileEvent] = [
            x for x in scheduled_now if x.type == EventKind.FILE  # type: ignore
        ]

        scheduled_now_webstream: List[WebStreamEvent] = [
            x  # type: ignore
            for x in scheduled_now
            if x.type == EventKind.WEB_STREAM_OUTPUT_START
        ]

        schedule_ids: Set[int] = {x.row_id for x in scheduled_now_files}

        row_id_map: Dict[int, FileEvent] = {}
        liq_queue_ids: Set[int] = set()
        for queue_item in self.liq_queue_tracker.values():
            if queue_item is not None and not queue_item.ended():
                liq_queue_ids.add(queue_item.row_id)
                row_id_map[queue_item.row_id] = queue_item

        to_be_removed: Set[int] = set()
        to_be_added: Set[int] = set()

        # Iterate over the new files, and compare them to currently scheduled
        # tracks. If already in liquidsoap queue still need to make sure they don't
        # have different attributes. Ff replay gain changes, it shouldn't change the
        # amplification of the currently playing song
        for item in scheduled_now_files:
            if item.row_id in row_id_map:
                queue_item = row_id_map[item.row_id]

                if not (
                    queue_item.start == item.start
                    and queue_item.end == item.end
                    and queue_item.row_id == item.row_id
                ):
                    # need to re-add
                    logger.info("Track %s found to have new attr.", item)
                    to_be_removed.add(item.row_id)
                    to_be_added.add(item.row_id)

        to_be_removed.update(liq_queue_ids - schedule_ids)
        to_be_added.update(schedule_ids - liq_queue_ids)

        if to_be_removed:
            logger.info("Need to remove items from Liquidsoap: %s", to_be_removed)

            # remove files from Liquidsoap's queue
            for queue_id, queue_item in self.liq_queue_tracker.items():
                if queue_item is not None and queue_item.row_id in to_be_removed:
                    self.stop(queue_id)

        if to_be_added:
            logger.info("Need to add items to Liquidsoap *now*: %s", to_be_added)

            for item in scheduled_now_files:
                if item.row_id in to_be_added:
                    self.modify_cue_point(item)
                    self.play(item)

        # handle webstreams
        current_stream_id = self.telnet_liquidsoap.get_current_stream_id()
        if current_stream_id is None:
            current_stream_id = "-1"

        logger.debug("scheduled now webstream: %s", scheduled_now_webstream)
        if scheduled_now_webstream:
            if int(current_stream_id) != int(scheduled_now_webstream[0].row_id):
                self.play(scheduled_now_webstream[0])
        elif current_stream_id != "-1":
            # something is playing and it shouldn't be.
            self.telnet_liquidsoap.stop_web_stream_buffer()
            self.telnet_liquidsoap.stop_web_stream_output()

    def stop(self, queue_id: int) -> None:
        self.telnet_liquidsoap.queue_remove(queue_id)
        self.liq_queue_tracker[queue_id] = None

    def clear_queue_tracker(self) -> None:
        for queue_id in self.liq_queue_tracker:
            self.liq_queue_tracker[queue_id] = None

    def modify_cue_point(self, file_event: FileEvent) -> None:
        assert file_event.type == EventKind.FILE

        lateness = seconds_between(file_event.start, datetime.utcnow())

        if lateness > 0:
            logger.debug("media item was supposed to start %ss ago", lateness)
            cue_in_orig = timedelta(seconds=file_event.cue_in)
            file_event.cue_in = cue_in_orig.total_seconds() + lateness

    def clear_all_queues(self) -> None:
        self.telnet_liquidsoap.queue_clear_all()


class UnknownEvent(Exception):
    pass


class NoQueueAvailableException(Exception):
    pass
