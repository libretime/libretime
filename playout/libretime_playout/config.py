from pathlib import Path
from typing import List, Literal, Optional

from libretime_shared.config import (
    BaseConfig,
    GeneralConfig,
    RabbitMQConfig,
    StreamConfig,
)
from pydantic import BaseModel, model_validator

CACHE_DIR = Path.cwd() / "scheduler"
RECORD_DIR = Path.cwd() / "recorder"

PUSH_INTERVAL: float = 2.0
POLL_INTERVAL: float = 400.0


class PlayoutConfig(BaseModel):
    liquidsoap_host: str = "localhost"
    liquidsoap_port: int = 1234

    cache_ahead_hours: int = 3

    record_file_format: Literal["mp3", "ogg"] = "ogg"  # record_file_type
    record_bitrate: int = 256
    record_samplerate: int = 44100
    record_channels: int = 2
    record_sample_size: int = 16


class LiquidsoapConfig(BaseModel):
    server_listen_address: str = "127.0.0.1"
    server_listen_port: int = 1234

    harbor_listen_address: List[str] = ["0.0.0.0"]

    harbor_ssl_certificate: Optional[str] = None
    harbor_ssl_private_key: Optional[str] = None
    harbor_ssl_password: Optional[str] = None

    @model_validator(mode="after")
    def _validate_harbor_ssl(self):
        if (
            self.harbor_ssl_certificate is not None
            and self.harbor_ssl_private_key is None
        ):
            raise ValueError("missing 'harbor_ssl_private_key' value")

        if (
            self.harbor_ssl_certificate is None
            and self.harbor_ssl_private_key is not None
        ):
            raise ValueError("missing 'harbor_ssl_certificate' value")

        return self


class Config(BaseConfig):
    general: GeneralConfig
    rabbitmq: RabbitMQConfig = RabbitMQConfig()
    playout: PlayoutConfig = PlayoutConfig()
    liquidsoap: LiquidsoapConfig = LiquidsoapConfig()
    stream: StreamConfig = StreamConfig()
