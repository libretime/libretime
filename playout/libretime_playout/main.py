"""
Python part of radio playout (pypo)
"""

import logging
import os
import sys
import time
from datetime import datetime
from pathlib import Path
from queue import Queue
from typing import Any, Dict, Optional

import click
import requests
from libretime_api_client.v1 import ApiClient as LegacyClient
from libretime_api_client.v2 import ApiClient
from libretime_shared.cli import cli_config_options, cli_logging_options
from libretime_shared.config import DEFAULT_ENV_PREFIX
from libretime_shared.logging import setup_logger

from . import PACKAGE, VERSION
from .config import CACHE_DIR, RECORD_DIR, Config
from .history.stats import StatsCollectorThread
from .liquidsoap.client import LiquidsoapClient
from .liquidsoap.version import LIQUIDSOAP_MIN_VERSION
from .message_handler import MessageListener
from .player.events import Events, FileEvents
from .player.fetch import PypoFetch
from .player.file import PypoFile
from .player.liquidsoap import Liquidsoap
from .player.push import PypoPush

logger = logging.getLogger(__name__)

for module in ("amqp",):
    logging.getLogger(module).setLevel(logging.INFO)
    logging.getLogger(module).propagate = False


def wait_for_legacy(legacy_client: LegacyClient) -> None:
    while legacy_client.version() == -1:
        time.sleep(2)

    success = False
    while not success:
        try:
            legacy_client.register_component("pypo")
            success = True
        except (
            requests.exceptions.ConnectionError,
            requests.exceptions.HTTPError,
            requests.exceptions.Timeout,
        ) as exception:
            logger.exception(exception)
            time.sleep(10)


def wait_for_liquidsoap(liq_client: LiquidsoapClient) -> None:
    logger.debug("Checking if Liquidsoap is running")
    liq_version = liq_client.wait_for_version()
    if not LIQUIDSOAP_MIN_VERSION <= liq_version:
        raise RuntimeError(f"Invalid liquidsoap version {liq_version}")


@click.command(context_settings={"auto_envvar_prefix": DEFAULT_ENV_PREFIX})
@cli_logging_options()
@cli_config_options()
def cli(
    log_level: str,
    log_filepath: Optional[Path],
    config_filepath: Optional[Path],
) -> None:
    """
    Run playout.
    """
    setup_logger(log_level, log_filepath)
    config = Config(config_filepath)

    if "SENTRY_DSN" in os.environ:
        logger.info("installing sentry")
        # pylint: disable=import-outside-toplevel
        import sentry_sdk

        sentry_sdk.init(
            traces_sample_rate=1.0,
            release=f"{PACKAGE}@{VERSION}",
        )

    try:
        for dir_path in [CACHE_DIR, RECORD_DIR]:
            dir_path.mkdir(exist_ok=True)
    except OSError as exception:
        logger.error(exception)
        sys.exit(1)

    # Although all of our calculations are in UTC, it is useful to know what timezone
    # the local machine is, so that we have a reference for what time the actual
    # log entries were made
    logger.info("Timezone: %s", time.tzname)
    logger.info("UTC time: %s", datetime.utcnow())

    api_client = ApiClient(
        base_url=config.general.public_url,
        api_key=config.general.api_key,
    )

    legacy_client = LegacyClient(
        base_url=config.general.public_url,
        api_key=config.general.api_key,
    )
    wait_for_legacy(legacy_client)

    liq_client = LiquidsoapClient(
        host=config.playout.liquidsoap_host,
        port=config.playout.liquidsoap_port,
    )
    wait_for_liquidsoap(liq_client)

    fetch_queue: "Queue[Dict[str, Any]]" = Queue()
    push_queue: "Queue[Events]" = Queue()
    # This queue is shared between pypo-fetch and pypo-file, where pypo-file
    # is the consumer. Pypo-fetch will send every schedule it gets to pypo-file
    # and pypo will parse this schedule to determine which file has the highest
    # priority, and retrieve it.
    file_queue: "Queue[FileEvents]" = Queue()

    liquidsoap = Liquidsoap(liq_client)

    PypoFile(file_queue, api_client).start()

    PypoFetch(
        fetch_queue,
        push_queue,
        file_queue,
        liq_client,
        liquidsoap,
        config,
        api_client,
        legacy_client,
    ).start()

    PypoPush(push_queue, liquidsoap, config).start()

    StatsCollectorThread(config, legacy_client).start()

    message_listener = MessageListener(config, fetch_queue)
    message_listener.run_forever()
