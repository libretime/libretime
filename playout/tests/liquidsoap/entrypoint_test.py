from pathlib import Path
from subprocess import CalledProcessError, check_call, check_output
from unittest import mock

import pytest

from libretime_playout.config import Config
from libretime_playout.liquidsoap.entrypoint import generate_entrypoint
from libretime_playout.liquidsoap.models import Info, StreamPreferences
from libretime_playout.liquidsoap.version import get_liquidsoap_version

from .conftest import LIQ_VERSION
from .fixtures import TEST_STREAM_CONFIGS, make_config_with_stream


@pytest.mark.parametrize(
    "version",
    [pytest.param((1, 4, 4), id="1.4")],
)
@pytest.mark.parametrize(
    "stream_config",
    TEST_STREAM_CONFIGS,
)
def test_generate_entrypoint(
    stream_config: Config,
    stream_preferences: StreamPreferences,
    version,
    snapshot,
):
    with mock.patch(
        "libretime_playout.liquidsoap.entrypoint.here",
        Path("/fake"),
    ):
        found = generate_entrypoint(
            log_filepath=Path("/var/log/radio.log"),
            config=stream_config,
            preferences=stream_preferences,
            info=Info(
                station_name="LibreTime",
            ),
            version=version,
        )

    assert found == snapshot


@pytest.mark.skipif(
    LIQ_VERSION >= (2, 0, 0),
    reason="unsupported liquidsoap >= 2.0.0",
)
@pytest.mark.parametrize(
    "stream_config",
    TEST_STREAM_CONFIGS,
)
def test_liquidsoap_syntax(
    tmp_path: Path,
    stream_config: Config,
    stream_preferences: StreamPreferences,
):
    entrypoint_filepath = tmp_path / "radio.liq"
    log_filepath = tmp_path / "radio.log"

    entrypoint_filepath.write_text(
        generate_entrypoint(
            log_filepath=log_filepath,
            config=stream_config,
            preferences=stream_preferences,
            info=Info(
                station_name="LibreTime",
            ),
            version=get_liquidsoap_version(),
        ),
        encoding="utf-8",
    )

    check_call(["liquidsoap", "--check", str(entrypoint_filepath)])


@pytest.mark.skipif(
    LIQ_VERSION >= (2, 0, 0),
    reason="unsupported liquidsoap >= 2.0.0",
)
def test_liquidsoap_unsupported_output_aac(
    tmp_path: Path,
    stream_preferences: StreamPreferences,
):
    entrypoint_filepath = tmp_path / "radio.liq"
    log_filepath = tmp_path / "radio.log"

    entrypoint_filepath.write_text(
        generate_entrypoint(
            log_filepath=log_filepath,
            config=make_config_with_stream(
                outputs={
                    "icecast": [
                        {
                            "enabled": True,
                            "mount": "main.aac",
                            "source_password": "hackme",
                            "audio": {"format": "aac", "bitrate": 128},
                        }
                    ]
                }
            ),
            preferences=stream_preferences,
            info=Info(
                station_name="LibreTime",
            ),
            version=get_liquidsoap_version(),
        ),
        encoding="utf-8",
    )

    with pytest.raises(CalledProcessError) as exception:
        check_output(["liquidsoap", "--check", str(entrypoint_filepath)])
    assert b"You must be missing an optional dependency." in exception.value.stdout
