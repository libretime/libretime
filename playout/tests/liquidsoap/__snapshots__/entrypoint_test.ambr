# serializer version: 1
# name: test_generate_entrypoint[stream_config0-1.4]
  '''
  # THIS FILE IS AUTO GENERATED. PLEASE DO NOT EDIT!
  ###########################################################
  # The ignore() lines are to squash unused variable warnings
  
  # Inputs
  input_main_mount = "main"
  input_main_port = 8001
  input_main_secure = false
  input_show_mount = "show"
  input_show_port = 8002
  input_show_secure = false
  
  # Settings
  set("log.file.path", "/var/log/radio.log")
  
  set("server.telnet", true)
  set("server.telnet.bind_addr", "127.0.0.1")
  set("server.telnet.port", 1234)
  
  set("harbor.bind_addrs", ["0.0.0.0"])
  
  station_name = interactive.string("station_name", "LibreTime")
  
  message_offline = interactive.string("message_offline", "LibreTime - offline")
  message_format = interactive.string("message_format", "0")
  input_fade_transition = interactive.float("input_fade_transition", 0.0)
  
  %include "/fake/1.4/ls_script.liq"
  
  
  
  
  
  gateway("started")
  
  '''
# ---
# name: test_generate_entrypoint[stream_config1-1.4]
  '''
  # THIS FILE IS AUTO GENERATED. PLEASE DO NOT EDIT!
  ###########################################################
  # The ignore() lines are to squash unused variable warnings
  
  # Inputs
  input_main_mount = "main"
  input_main_port = 8001
  input_main_secure = false
  input_show_mount = "show"
  input_show_port = 8002
  input_show_secure = false
  
  # Settings
  set("log.file.path", "/var/log/radio.log")
  
  set("server.telnet", true)
  set("server.telnet.bind_addr", "127.0.0.1")
  set("server.telnet.port", 1234)
  
  set("harbor.bind_addrs", ["0.0.0.0"])
  
  set("harbor.ssl.certificate", "/fake/ssl.cert")
  set("harbor.ssl.private_key", "/fake/ssl.key")
  
  station_name = interactive.string("station_name", "LibreTime")
  
  message_offline = interactive.string("message_offline", "LibreTime - offline")
  message_format = interactive.string("message_format", "0")
  input_fade_transition = interactive.float("input_fade_transition", 0.0)
  
  %include "/fake/1.4/ls_script.liq"
  
  
  
  
  
  gateway("started")
  
  '''
# ---
# name: test_generate_entrypoint[stream_config2-1.4]
  '''
  # THIS FILE IS AUTO GENERATED. PLEASE DO NOT EDIT!
  ###########################################################
  # The ignore() lines are to squash unused variable warnings
  
  # Inputs
  input_main_mount = "main"
  input_main_port = 8001
  input_main_secure = false
  input_show_mount = "show"
  input_show_port = 8002
  input_show_secure = false
  
  # Settings
  set("log.file.path", "/var/log/radio.log")
  
  set("server.telnet", true)
  set("server.telnet.bind_addr", "127.0.0.1")
  set("server.telnet.port", 1234)
  
  set("harbor.bind_addrs", ["0.0.0.0"])
  
  station_name = interactive.string("station_name", "LibreTime")
  
  message_offline = interactive.string("message_offline", "LibreTime - offline")
  message_format = interactive.string("message_format", "0")
  input_fade_transition = interactive.float("input_fade_transition", 0.0)
  
  %include "/fake/1.4/ls_script.liq"
  
  
  
  # icecast:1
  output_icecast_1_source = s
  # Disable ogg metadata
  output_icecast_1_source = add(normalize=false, [amplify(0.00001, noise()), output_icecast_1_source])
  output.icecast(
    id="icecast:1",
    host="localhost",
    port=8000,
    mount="main",
    user="source",
    password="hackme",
    fallible = true,
    name="LibreTime!",
    description="LibreTime Radio! Stream #1",
    genre="various",
    url="https://libretime.org",
    on_connect=make_ouput_on_connect_handler("1"),
    on_error=make_ouput_on_error_handler("1"),
    %ogg(%vorbis(
      quality=0.8,
      channels=2
    )),
    output_icecast_1_source,
  )
  
  
  
  gateway("started")
  
  '''
# ---
# name: test_generate_entrypoint[stream_config3-1.4]
  '''
  # THIS FILE IS AUTO GENERATED. PLEASE DO NOT EDIT!
  ###########################################################
  # The ignore() lines are to squash unused variable warnings
  
  # Inputs
  input_main_mount = "main"
  input_main_port = 8001
  input_main_secure = false
  input_show_mount = "show"
  input_show_port = 8002
  input_show_secure = false
  
  # Settings
  set("log.file.path", "/var/log/radio.log")
  
  set("server.telnet", true)
  set("server.telnet.bind_addr", "127.0.0.1")
  set("server.telnet.port", 1234)
  
  set("harbor.bind_addrs", ["0.0.0.0"])
  
  station_name = interactive.string("station_name", "LibreTime")
  
  message_offline = interactive.string("message_offline", "LibreTime - offline")
  message_format = interactive.string("message_format", "0")
  input_fade_transition = interactive.float("input_fade_transition", 0.0)
  
  %include "/fake/1.4/ls_script.liq"
  
  
  
  # shoutcast:1
  output_shoutcast_1_source = s
  output.shoutcast(
    id="shoutcast:1",
    host="localhost",
    port=8000,
    user="source",
    password="hackme",
    fallible = true,
    name="",
    genre="various",
    url="https://libretime.org",
    on_connect=make_ouput_on_connect_handler("1"),
    on_error=make_ouput_on_error_handler("1"),
    %mp3(
      bitrate=256,
      stereo=true
    ),
    output_shoutcast_1_source,
  )
  
  
  
  gateway("started")
  
  '''
# ---
# name: test_generate_entrypoint[stream_config4-1.4]
  '''
  # THIS FILE IS AUTO GENERATED. PLEASE DO NOT EDIT!
  ###########################################################
  # The ignore() lines are to squash unused variable warnings
  
  # Inputs
  input_main_mount = "main"
  input_main_port = 8001
  input_main_secure = false
  input_show_mount = "show"
  input_show_port = 8002
  input_show_secure = false
  
  # Settings
  set("log.file.path", "/var/log/radio.log")
  
  set("server.telnet", true)
  set("server.telnet.bind_addr", "127.0.0.1")
  set("server.telnet.port", 1234)
  
  set("harbor.bind_addrs", ["0.0.0.0"])
  
  station_name = interactive.string("station_name", "LibreTime")
  
  message_offline = interactive.string("message_offline", "LibreTime - offline")
  message_format = interactive.string("message_format", "0")
  input_fade_transition = interactive.float("input_fade_transition", 0.0)
  
  %include "/fake/1.4/ls_script.liq"
  
  
  
  # icecast:1
  output_icecast_1_source = s
  # Disable ogg metadata
  output_icecast_1_source = add(normalize=false, [amplify(0.00001, noise()), output_icecast_1_source])
  output.icecast(
    id="icecast:1",
    host="localhost",
    port=8000,
    mount="main",
    user="source",
    password="hackme",
    fallible = true,
    name="",
    description="",
    genre="",
    url="",
    on_connect=make_ouput_on_connect_handler("1"),
    on_error=make_ouput_on_error_handler("1"),
    %ogg(%vorbis(
      quality=0.8,
      channels=2
    )),
    output_icecast_1_source,
  )
  
  # shoutcast:2
  output_shoutcast_2_source = s
  output.shoutcast(
    id="shoutcast:2",
    host="localhost",
    port=8000,
    user="source",
    password="hackme",
    fallible = true,
    name="",
    genre="",
    url="",
    on_connect=make_ouput_on_connect_handler("2"),
    on_error=make_ouput_on_error_handler("2"),
    %mp3(
      bitrate=256,
      stereo=true
    ),
    output_shoutcast_2_source,
  )
  
  
  
  gateway("started")
  
  '''
# ---
# name: test_generate_entrypoint[stream_config5-1.4]
  '''
  # THIS FILE IS AUTO GENERATED. PLEASE DO NOT EDIT!
  ###########################################################
  # The ignore() lines are to squash unused variable warnings
  
  # Inputs
  input_main_mount = "main"
  input_main_port = 8001
  input_main_secure = false
  input_show_mount = "show"
  input_show_port = 8002
  input_show_secure = false
  
  # Settings
  set("log.file.path", "/var/log/radio.log")
  
  set("server.telnet", true)
  set("server.telnet.bind_addr", "127.0.0.1")
  set("server.telnet.port", 1234)
  
  set("harbor.bind_addrs", ["0.0.0.0"])
  
  station_name = interactive.string("station_name", "LibreTime")
  
  message_offline = interactive.string("message_offline", "LibreTime - offline")
  message_format = interactive.string("message_format", "0")
  input_fade_transition = interactive.float("input_fade_transition", 0.0)
  
  %include "/fake/1.4/ls_script.liq"
  
  
  
  # pulseaudio:1
  %ifndef output.pulseaudio
  log("output.pulseaudio is not defined!")
  %endif
  %ifdef output.pulseaudio
  output.pulseaudio(
    id="pulseaudio:1",
    s
  )
  %endif
  
  
  
  gateway("started")
  
  '''
# ---
# name: test_generate_entrypoint[stream_config6-1.4]
  '''
  # THIS FILE IS AUTO GENERATED. PLEASE DO NOT EDIT!
  ###########################################################
  # The ignore() lines are to squash unused variable warnings
  
  # Inputs
  input_main_mount = "main"
  input_main_port = 8001
  input_main_secure = false
  input_show_mount = "show"
  input_show_port = 8002
  input_show_secure = false
  
  # Settings
  set("log.file.path", "/var/log/radio.log")
  
  set("server.telnet", true)
  set("server.telnet.bind_addr", "127.0.0.1")
  set("server.telnet.port", 1234)
  
  set("harbor.bind_addrs", ["0.0.0.0"])
  
  station_name = interactive.string("station_name", "LibreTime")
  
  message_offline = interactive.string("message_offline", "LibreTime - offline")
  message_format = interactive.string("message_format", "0")
  input_fade_transition = interactive.float("input_fade_transition", 0.0)
  
  %include "/fake/1.4/ls_script.liq"
  
  
  
  # pulseaudio:1
  %ifndef output.pulseaudio
  log("output.pulseaudio is not defined!")
  %endif
  %ifdef output.pulseaudio
  output.pulseaudio(
    id="pulseaudio:1",
    device="alsa_output.pci-0000_00_sink",
    s
  )
  %endif
  
  
  
  gateway("started")
  
  '''
# ---
# name: test_generate_entrypoint[stream_config7-1.4]
  '''
  # THIS FILE IS AUTO GENERATED. PLEASE DO NOT EDIT!
  ###########################################################
  # The ignore() lines are to squash unused variable warnings
  
  # Inputs
  input_main_mount = "main"
  input_main_port = 8001
  input_main_secure = false
  input_show_mount = "show"
  input_show_port = 8002
  input_show_secure = false
  
  # Settings
  set("log.file.path", "/var/log/radio.log")
  
  set("server.telnet", true)
  set("server.telnet.bind_addr", "127.0.0.1")
  set("server.telnet.port", 1234)
  
  set("harbor.bind_addrs", ["0.0.0.0"])
  
  station_name = interactive.string("station_name", "LibreTime")
  
  message_offline = interactive.string("message_offline", "LibreTime - offline")
  message_format = interactive.string("message_format", "0")
  input_fade_transition = interactive.float("input_fade_transition", 0.0)
  
  %include "/fake/1.4/ls_script.liq"
  
  
  
  
  
  gateway("started")
  
  '''
# ---
